aws-codeguruprofiler-profilinggroup/src/test/java/software/amazon/codeguruprofiler/profilinggroup/CreateHandlerTest.java

```diff
@@ -43,83 +53,226 @@
-                .when(proxy).injectCredentialsAndInvokeV2(any(), any());
+    @Nested
+    class WithPermissions {
+
+        @BeforeEach
+        public void setup() {
+            doReturn(CreateProfilingGroupResponse.builder().build())
+                    .when(proxy).injectCredentialsAndInvokeV2(any(CreateProfilingGroupRequest.class), any());
+        }
+
+        @Nested
+        class NoPrincipals {
+
+            @Test
+            public void testNullPermissions() {
+                ResourceModel model = newResourceModel(null);
+                request = makeRequest(model);
+                assertResponse(subject.handleRequest(proxy, request, null, logger));
+            }
+
+            @Test
+            public void testNullPrincipals() {
+                ResourceModel model = newResourceModel(AgentPermissions.builder().principals(null).build());
+                request = makeRequest(model);
+                assertResponse(subject.handleRequest(proxy, request, null, logger));
+            }
+
+            @AfterEach
+            public void testCorrectOperationIsCalled() {
+                verify(proxy, times(1)).injectCredentialsAndInvokeV2(eq(createPgRequest), any());
+                verifyNoMoreInteractions(proxy);
+            }
+        }
+
+        @Nested
+        class WithPrincipals {
+
+            @BeforeEach
+            public void setup() {
+                ResourceModel model = newResourceModel(AgentPermissions.builder().principals(Arrays.asList("a", "bc")).build());
+                request = makeRequest(model);
+            }
+
+            @Test
+            public void testSuccess() {
+                assertResponse(subject.handleRequest(proxy, request, null, logger));
+                verify(proxy, times(1)).injectCredentialsAndInvokeV2(eq(createPgRequest), any());
+                verify(proxy, times(1)).injectCredentialsAndInvokeV2(any(PutPermissionRequest.class), any());
+                verifyNoMoreInteractions(proxy);
+            }
+
+            @Test
+            public void testPutPermissionsFails() {
+                doThrow(ConflictException.builder().build())
+                        .when(proxy).injectCredentialsAndInvokeV2(any(PutPermissionRequest.class), any());
+
+                CfnAlreadyExistsException exception = assertThrows(CfnAlreadyExistsException.class,
+                        () -> subject.handleRequest(proxy, request, null, logger));
+
+                assertThat(exception).hasCauseExactlyInstanceOf(ConflictException.class);
+                assertThat(exception).hasNoSuppressedExceptions();
 
-        assertThrows(CfnServiceLimitExceededException.class, () ->
-                new CreateHandler().handleRequest(proxy, request, null, logger));
+                verify(proxy, times(1)).injectCredentialsAndInvokeV2(eq(createPgRequest), any());
+                verify(proxy, times(1)).injectCredentialsAndInvokeV2(any(PutPermissionRequest.class), any());
```

Comment:
On this one, I'm not sure it makes sense to validate the requests -- we know that the request needs to happen because we observed the exception that is its result.

(Same for the next test)

Reply:
Okay, I will remove them.

(Same for the next test)